### DISCUSSION TITLE
Populate tests from module-level fixture

### DISCUSSION THREAD

**Discussion by AndySemenov** [pytest-dev/pytest#9945] [Q&A] [answered]
A fixture (scope='module') is bound to a test function as a parameter. That function pulls some data from a remote database or from a local file snapshot (depends on command line parameters and/or environment variables that could override) which defines some rules according to which the testcase should generate values to run through the business logic being tested and also populates some variables that describe what items were processed, which of them generate a single test function run and so on. So the fixture itself is the function that defines the parameters for test function parametrization. But according to the code `pytest_generate_tests(metafunc)` is the final signature for the hook and thus cannot acquire other fixtures as additional parameters. How could the module-level fixture results be used in a test function parametrization then?

All the examples in the documentation just define some static switching logic (like class A for this case and class B for another and here we go on with parametrization on those two).

**Comment by nicoddemus** [comment, marked as answer]
Hi @AndySemenov,

A normal pytest session is designed  to execute two separate stages:

1. Collection, where we find out which tests exist (along with other `conftest.py` files discovered while we search the file system). At this stage `pytest_generate_tests` kicks in for each module it is defined, and can generate more tests. At this point we also discover fixtures, their scopes, and their visibility to each test (no fixture code is executed yet).
2. Test run: at this stage, for each collected test, we execute it, which happens in 3 steps: `setup`, `call`, and `teardown`. During `setup`, each fixture requested by the test is executed, with the return value injected into the test function.

Under this design, is not possible for fixture code (which executes in stage 2) to generate more tests. The only solution is to decouple the "test generation" code from your fixture, and execute it during `pytest_generate_tests`.

Hope the general overview helps!

**Reply by AndySemenov** [reply]
But, as I've said, `pytest_generate_tests` does not have an access to `pytestconfig` fixture which is essential for the code that generates tests since it's the only fixture one could access command line parameters used in this certain run.

**Reply by nicoddemus** [reply]
Oh you never mentioned the fixture you wanted was `pytestconf`.

You can access that thru [`metafunc.config`](https://docs.pytest.org/en/stable/reference/reference.html#pytest.Metafunc.config) in `pytest_generate_tests`. 👍

**Reply by AndySemenov** [reply]
I'll mark top-level message as the answer, though, the reference to `metafunc.config` is actually what I needed. Thanks, missed that in the API docs.
